src/lib/controllers/history-controller.js

```diff
@@ -0,0 +1,45 @@
+export default class HistoryController {
+  /**
+   * Starts saving history for the alignment
+   * @param {Alignment} alignment
+   */
+  startTracking (alignment) {
+    this.alignment = alignment
+  }
+
+  /**
+   * Step back
+   *   if there is an active alignment group with more then 1 element, then execute undo inside an active alignment group
+   *   if there is an active alignment group with only one element, then we remove the group
+   *   if there is no active alignment group but there exists saved alignment groups, then we would activate previous group
+   */
+  undo () {
+    if (this.alignment.hasActiveAlignment && this.alignment.activeAlignmentGroup.groupLen > 1) {
+      return this.alignment.undoInActiveGroup()
+    }
+    if (this.alignment.hasActiveAlignment && this.alignment.activeAlignmentGroup.groupLen === 1) {
+      return this.alignment.undoActiveGroup()
+    }
+    if (!this.alignment.hasActiveAlignment && this.alignment.alignmentGroups.length > 0) {
+      return this.alignment.activateGroupByGroupIndex(this.alignment.alignmentGroups.length - 1)
+    }
+  }
+
+  /**
+   * Step forward
+   *   if there is an active alignment group and there are some steps that were undone, then execute redo inside active group
+   *   if there is an active alignment group and there are no undone steps, then we simply finish the group
+   *   if there is no active alignment group and there are some saved undone groups, then we would reactivate next group from the list
+   */
+  redo () {
+    if (this.alignment.hasActiveAlignment && !this.alignment.currentStepOnLastInActiveGroup) {
+      return this.alignment.redoInActiveGroup()
+    }
+    if (this.alignment.hasActiveAlignment && this.alignment.currentStepOnLastInActiveGroup) {

```

Comment:
There are several ways to do here:
1. (Done) we have history through groups - this way we could undo/redo inside an active group, we could undo/redo the whole group - that's why when the group is gone through with redo, we save it (without checking, were it previously saved or not)
2. We could add to steps - `save group`, and then we could revert it only if we have it in history

But anyway I believe we need the following states:
  - we have an active group and redo in it
  - we have an active group and save it with redo (the state between we finish the alignment group and start the new one)
  - we have no active group and activates previously undone

I realized the first variant - because it is easier without checking if the group was peviously saved or not.
Because we could start to undo on the moment group was already saved and no group was active, or on the moment group was not yet saved.
@balmas , what do you think? Should I add a step that group was saved or not and track it ? Or it is unuseful complexity?

Reply:
I'm sorry I still don't really follow. Using Redo to Save the group just doesn't make sense to me from a usability perspective.

(I realize that my comment above was missing an important word: "I don't think redo should *NOT* be possible if there are no steps to redo. saving and finishing the current group doesn't make sense to me.)

Reply:
ok, I think that our misunderstanding means that I should add a step for saving the group.
It would make the workflow more clear. Would do that